File: core/src/adictive/games/TiledMap.java
package adictive.games;

import com.badlogic.ashley.core.Entity;

import adictive.games.components.BlackHoleComponent;
import adictive.games.components.CoinComponent;
import adictive.games.components.SpikeComponent;
import adictive.games.components.WallComponent;
import adictive.games.components.WinComponent;
import adictive.games.systems.Reseteable;

public class TiledMap implements Reseteable {
    public Entity[][][] tiles;
    private static final int MAX_ENTITIES_ON_TILE = 4;

    public static final Class[] REGISTRY = new Class[6];

    public static final byte WALL = 1;
    public static final byte WIN = 2;
    public static final byte COIN = 3;
    public static final byte SPIKE = 4;
    public static final byte HOLE = 5;

    static {
        REGISTRY[WALL] = WallComponent.class;
        REGISTRY[WIN] = WinComponent.class;
        REGISTRY[COIN] = CoinComponent.class;
        REGISTRY[SPIKE] = SpikeComponent.class;
        REGISTRY[HOLE] = BlackHoleComponent.class;
    }

    public void init(int width, int height) {
        this.tiles = new Entity[width][height][REGISTRY.length];
    }

    /*
     * O(len(REGISTRY))
     */
    public void setEntity(int x, int y, Entity entity) {
        int entityConstant = getEntityConstant(entity);
        if (entityConstant >= 0) {
            tiles[x][y][entityConstant] = entity;
        }
    }

    public void removeEntity(int x, int y, Entity entity) {
        int entityConstant = getEntityConstant(entity);
        if (entityConstant >= 0) {
            tiles[x][y][entityConstant] = null;
        }
    }

    public Entity getEntity(int x, int y, byte entityType) {
        return this.tiles[x][y][entityType];
    }

    @SuppressWarnings("unchecked")
    private int getEntityConstant(Entity entity) {
        final int size = REGISTRY.length;
        for (int i = 0; i < size; i++) {
            if (REGISTRY[i] != null && entity.getComponent(REGISTRY[i]) != null) {
                return i;
            }
        }
        return -1;
    }

    public void reset() {
        for (int i = 0; i < tiles.length; i++) {
            for (int j = 0; j < tiles[i].length; j++) {
                for (int k = 0; k < MAX_ENTITIES_ON_TILE; k++) {
                    tiles[i][j][k] = null;
                }
            }
        }
    }
}

File: core/src/adictive/games/TiledMap.kt
package adictive.games

import adictive.games.components.*
import adictive.games.systems.Reseteable
import com.badlogic.ashley.core.Entity

class TiledMap : Reseteable {

    lateinit var tiles: Array<Array<Array<Entity?>>>

    fun init(width: Int, height: Int) {
        this.tiles = Array(width) { _ ->
            Array(height) { _ ->
                Array<Entity?>(REGISTRY.size) { null }
            }
        }
    }

    /*
     * O(len(REGISTRY))
     */
    fun setEntity(x: Int, y: Int, entity: Entity) {
        val entityConstant = getEntityConstant(entity)
        if (entityConstant >= 0) {
            tiles[x][y][entityConstant] = entity
        }
    }

    fun removeEntity(x: Int, y: Int, entity: Entity) {
        val entityConstant = getEntityConstant(entity)
        if (entityConstant >= 0) {
            tiles[x][y][entityConstant] = null
        }
    }

    fun getEntity(x: Int, y: Int, entityType: Int): Entity? {
        return this.tiles[x][y][entityType]
    }

    private fun getEntityConstant(entity: Entity): Int {
        val size = REGISTRY.size
        for (i in 0 until size) {
            if (entity.getComponent(REGISTRY[i]) != null) {
                return i
            }
        }
        return -1
    }

    override fun reset() {
        for (i in tiles.indices) {
            for (j in 0 until tiles[i].size) {
                for (k in 0 until MAX_ENTITIES_ON_TILE) {
                    tiles[i][j][k] = null
                }
            }
        }
    }

    companion object {
        const val MAX_ENTITIES_ON_TILE = 4

        const val WALL = 1
        const val WIN = 2
        const val COIN = 3
        const val SPIKE = 4
        const val HOLE = 5
        const val FLOOR = 6

        val REGISTRY = Array(7) {
            when (it) {
                WALL -> WallComponent::class.java
                WIN -> WinComponent::class.java
                COIN -> CoinComponent::class.java
                SPIKE -> SpikeComponent::class.java
                HOLE -> BlackHoleComponent::class.java
                FLOOR -> FloorEffectComponent::class.java
                else -> {
                    EmptyComponent::class.java
                }
            }
        }
    }
}
